python/src/iceberg/types.py

```diff
@@ -230,133 +191,260 @@ class LongType(Type):
     Iceberg are stored as UTC and include a date and a time of day with a timezone.
 
     Example:
-        >>> column_foo = TimestamptzType()
-        >>> isinstance(column_foo, TimestamptzType)
+        >>> column_foo = Timestamptz()
+        >>> isinstance(column_foo, Timestamptz)
         True
+        >>> Timestamptz("2017-11-16T14:31:08-08:00")
+        Timestamptz(value="2017-11-16T14:31:08-08:00")
     """
 
-    def __init__(self):
-        super().__init__("timestamptz", "TimestamptzType()", is_primitive=True)
+    def __init__(self, value: Optional[str] = None):
+        super().__init__(value)
 
 
-class StringType(Type):
+class String(PrimitiveType):
     """A String data type in Iceberg can be represented using an instance of this class. Strings in
     Iceberg are arbitrary-length character sequences and are encoded with UTF-8.
 
+    Args:
+        value: an optional argument for creating a literal without which instantiation represents just a type
+
     Example:
-        >>> column_foo = StringType()
-        >>> isinstance(column_foo, StringType)
+        >>> column_foo = String()
+        >>> isinstance(column_foo, String)
         True
+        >>> String("hello")
+        String(value="hello")
     """
 
-    def __init__(self):
-        super().__init__("string", "StringType()", is_primitive=True)
+    def __init__(self, value: Optional[str] = None):
+        super().__init__(value)
 
 
-class UUIDType(Type):
+class UUID(PrimitiveType):
     """A UUID data type in Iceberg can be represented using an instance of this class. UUIDs in
     Iceberg are universally unique identifiers.
 
     Example:
-        >>> column_foo = UUIDType()
-        >>> isinstance(column_foo, UUIDType)
+        >>> column_foo = UUID()
+        >>> isinstance(column_foo, UUID)
         True
+        >>> UUID("f79c3e09-677c-4bbd-a479-3f349cb785e7")
+        UUID(value="f79c3e09-677c-4bbd-a479-3f349cb785e7")
     """
 
-    def __init__(self):
-        super().__init__("uuid", "UUIDType()", is_primitive=True)
+    def __init__(self, value: Optional[str] = None):
+        super().__init__(value)
 
 
-class BinaryType(Type):
+class Binary(PrimitiveType):
     """A Binary data type in Iceberg can be represented using an instance of this class. Binarys in
     Iceberg are arbitrary-length byte arrays.
 
+    Args:
+        value: an optional argument for creating a literal without which instantiation represents just a type
+
     Example:
-        >>> column_foo = BinaryType()
-        >>> isinstance(column_foo, BinaryType)
+        >>> column_foo = Binary()
+        >>> isinstance(column_foo, Binary)
         True
+        >>> Binary(b'hello'))
+
     """
 
-    def __init__(self):
-        super().__init__("binary", "BinaryType()", is_primitive=True)
+    def __init__(self, value: Optional[bytes] = None):
+        super().__init__(value)
+
+
+class NestedField(object):
```

Comment:
Rearranging the classes appears to have introduced a lot more changes than needed. We ended up with primitives being diffed against the old nested types, and we don't get to see changes for these classes. I think making this a two-step change would allow us to see what changed much more easily. Could you move these back to the top, or move them and then base the PR on master after the move?